### PR TITLE
fix(client): explain the suffixed name when a bare claim is unavailable

### PR DESCRIPTION
**Pull request by Celant**
Prepares the client for infra #593 (OPE-225), which stops 409-ing a premium player whose chosen bare name is already held. After that change the API grants the suffixed form and returns **200** with `bareClaim: "unavailable"`.

**This must land before #593.** It is not a nice-to-have ordering preference — without it the deployed client is strictly *worse off* than the bug #593 fixes.

## Why

`UsernamePanel.handleSave` calls `window.location.reload()` on any 200 without inspecting the body. So once #593 ships to an unprepared client:

- premium player types `Ninja` → 200 → reload → modal reopens showing `Ninja.4471` → **no message at all**
- and because the fallback is a *real* rename, their 30-day cooldown is spent, locking them to a name they never chose

What that replaces is a 409 that said plainly "that username is already taken — try another", with no rename and no cooldown consumed. Trading a clear error for a silent month-long loss is a regression.

It is forward-compatible, so landing it first has no bad window in either direction: with #593 unmerged the API simply never sends the field and behaviour is byte-for-byte what it is today.

## Changes

1. **`bareClaim` added to `PutUsernameResponseSchema`** (`src/core/ApiSchemas.ts`) as a 3-value enum — `claimed` / `unavailable` / `not_eligible`. Three values rather than a boolean so callers don't have to re-derive eligibility from `usernameStatus` in order to avoid telling a free player that their perfectly ordinary suffixed rename "fell back". Zod strips unknown keys rather than erroring, which is why the field is silently dropped today and nothing looks broken.

2. **`UsernamePanel` explains the fallback before reloading** — naming what they asked for, what they actually got, and that the change used their rename. Awaited, so the reload cannot race the dialog away. Silent for `claimed` and `not_eligible`.

3. **The stale comment on `updateUsername`** (`src/client/Api.ts`) no longer claims both 409s are claim conflicts. The claim-conflict 409 is gone; the "name equals an existing public id" and "discriminator exhausted" ones remain and still map to `taken`.

## One deviation from the brief, and it matters

**`bareClaim` is `.optional()`, not required.**

The contract says the field is always present and never undefined — true, but only *after* #593 deploys. This PR ships first, and `updateUsername` reads the response with `safeParse`:

```ts
const parsed = PutUsernameResponseSchema.safeParse(await response.json());
if (!parsed.success) {
  return { ok: false, code: "failed" };
}
```

A required field would therefore make **every rename against the current API** fail validation and surface as a generic "failed" — breaking renames outright for everyone in exactly the window this PR exists to protect. Optional is what makes it safe to land first, and `undefined` is read as "the API predates this field, say nothing". The first test in the new file pins that.

## The date

The message names the date they can next rename, taken from `nextUsernameChangeAt` in the same response rather than hardcoding 30 days — the server is authoritative and the panel already formats dates this way for the grace warning. `nextUsernameChangeAt` is nullable, so there is a second key without the clause rather than a message with a blank date in it.

## Testing

`tests/client/UsernameBareClaim.test.ts` — 10 cases:

- schema parses a response **without** the field (the pre-#593 case that must not break), keeps all three values once sent, and rejects a value outside the enum rather than passing it through
- the panel explains the fallback and names both the requested and granted names before reloading
- it uses the date when supplied and drops the clause rather than rendering a blank one when null
- it says nothing for `claimed`, `not_eligible`, or an absent field — each asserting the reload still happens
- the surviving 409s still surface as an inline error with no dialog and no reload

Full suite: 347 files, 4175 tests. `tsc --noEmit`, `prettier --check .`, `npm run lint` clean. (`tests/client/InventoryModal.test.ts` intermittently fails under full-suite parallel load on this machine and passes when run alone — the known flake, unrelated to this change.)

Deliberately **not** the OPE-222 claim prompt. This is only the explanation that stops a silent loss; resolve-before-submit comes with OPE-222.

### CODE REVIEW

**Review by Celant — COMMENTED** [review]
## Review — OPE-265, explain the suffixed name when a bare claim is unavailable

**The `.optional()` inversion is right, and the case for it is stronger than the PR states.** No must-fix findings. Two things worth changing before merge, both one-liners, and one note that will otherwise trip up whoever reads this diff locally.

Verified in a clean worktree with a real `npm ci --ignore-scripts`: **10/10 new tests pass, `tsc --noEmit` clean.** Full-suite state is at the bottom — my first run showed two failures and neither turned out to be this PR, but I chased both rather than calling them flake.

---

### The load-bearing question: is `bareClaim` right as `.optional()`?

Yes. I traced the failure path rather than taking the reasoning on trust, and it is worse than "surfaces as a generic failed".

`updateUsername` parses with `safeParse` and maps failure straight to `{ ok: false, code: "failed" }` (`src/client/Api.ts:387-391`), which `UsernamePanel.errorMessage` renders as `username_error_failed`. So against today's API — which does not send the field, because #593 has not merged — a **required** `bareClaim` would fail validation on every single rename.

The part worth spelling out: **the rename has already committed server-side by then.** The client is parsing a 200. So a required field would not merely show a spurious error — it would report failure for a rename that *succeeded*, spend the player's 30-day cooldown, and leave the modal reopening on a name they did not choose with an error saying it did not work. That is strictly worse than the bug this PR exists to fix, and it would hit **100% of renames**, not an edge case.

**Both directions check out.** Field absent → `undefined` → `warnBareClaimUnavailable` returns early → behaviour identical to today, pinned behaviourally at `tests/client/UsernameBareClaim.test.ts:195-203`. Field present → handled, all three values. And if #593 were ever rolled back, the client silently reverts to saying nothing rather than breaking.

**Does the first test pin `undefined` or merely mention it? — It pins it, verified by mutation.** Changing `BareClaimSchema.optional()` to `BareClaimSchema` fails exactly one test:

```
× parses a response from an API that does not send the field
  Tests  1 failed | 9 passed (10)
```

Worth knowing precisely where that guard lives: only the schema-level test at `:45-49` catches it. The behavioural test at `:195` mocks `updateUsername` wholesale, so it never reaches `safeParse` and would not have caught a required-field regression. One test, at the right level, is enough — but it is one test, not two.

---

### Worth considering

**1. The enum is strict, which reintroduces this same hazard one release later — `src/core/ApiSchemas.ts:302`**

`bareClaim: BareClaimSchema.optional()` accepts *absent*, but rejects any value outside the three. The test at `:61-66` pins that deliberately ("rejects a value outside the enum rather than passing it through"), so this was a choice rather than an oversight.

But the justification for `.optional()` is that this client is deployed independently of, and often behind, the API. That skew does not end when #593 merges — it recurs on every future API change. If infra ever adds a fourth `bareClaim` value, every un-updated client fails `safeParse` → `"failed"` → rename succeeded, cooldown spent, generic error. **The identical failure mode this PR exists to prevent, one release later.**

The cost asymmetry is what makes this worth changing. `warnBareClaimUnavailable` acts only on `=== "unavailable"`, so an unknown value treated as `undefined` degrades to "say nothing" — which is exactly the documented, intended behaviour for an older client. Rejecting degrades to "break the rename UX for everyone".

```ts
bareClaim: BareClaimSchema.optional().catch(undefined),
```

zod is `^4.4.2`, so `.catch()` is available. The test at `:61-66` inverts to assert `success === true` with `bareClaim === undefined`, which is a more useful thing to pin than the rejection. If strictness is deliberate for a reason I am not seeing, the PR should say why absent and unknown are treated differently, because the reasoning that justifies one appears to justify the other.

**2. The awaited dialog is correct, but no test holds it that way — mutation-verified**

`UsernamePanel.ts:127-131` says "Awaited so the reload cannot race the dialog away", and the code does exactly that. But replacing

```ts
await this.warnBareClaimUnavailable(name, result.data);
```

with `void this.warnBareClaimUnavailable(...)` — precisely the regression that comment guards against — **passes all 10 tests**. Both `showInGameAlert` and `reload` are mocks, so both still record a call; nothing asserts the order between them, and in a real browser the reload would tear the dialog away before the player read it.

One assertion in the first test closes it:

```ts
expect(showInGameAlert.mock.invocationCallOrder[0])
  .toBeLessThan(reload.mock.invocationCallOrder[0]);
```

This is the failure mode you flagged as one unit tests tend to miss, and it is: the code is right, the suite does not know it is right.

**3. A hang in the alert now strands a committed rename**

`showDialog` (`src/client/InGameModal.ts:16-32`) resolves only from a `confirm` or `cancel` event and never rejects, so if neither fires the promise never settles: `reload()` never runs and `busy` stays `true`.

This is not new — the `showInGameConfirm` earlier in `handleSave` has the identical shape. The asymmetry is the consequence. A hang in the confirm leaves nothing done; a hang in the alert leaves the rename **committed and the cooldown spent**, with the form locked and the old name still displayed. Low probability and the same mechanism as existing code, so I would not change anything for it — worth knowing rather than fixing.

**4. Read the GitHub diff, not a local `origin/main..pr5213` — the branch is one commit behind**

Flagging because I nearly filed it as a finding. The branch is based on `900bde7e1`; `origin/main` is one ahead at `ae63bafe2` (#5206). So a local `git diff origin/main..pr5213` renders #5206's `Main.ts` change as a **38-line deletion** — it reads exactly like an unrelated revert of a fix that landed one commit ago.

It is not. Commit `6da140213` touches only the five intended files, and I confirmed the branch **merges clean** against current `origin/main`. No action needed; a rebase would remove the illusion if anyone else is reviewing locally.

---

### On leaving OPE-266 out — agreed, and it is the right call

Fixing the `isVerifiedUsername` / `accountVerifiedName` disagreement means deciding whether a claimless premium player gets a verified badge, and that is a product question rather than a mechanical one. It also touches `PlayerName.ts`, `UsernameInput.ts` and the badge render sites, none of which this PR goes near. Bundling it would widen the review surface of the one PR #593 is queued behind, for no throughput gain.

The one thing I would hold onto: OPE-266 becomes **reachable the moment #593 merges**, not when the client work lands. Until then a fallback player is granted `verified` server-side and rendered without the badge on friends lists, clan lists, leaderboards and profiles. Worth not letting it drift more than a release behind.

---

### Also verified

- **The nullable-date path renders sensibly**, and is pinned at `:161-173`. It produces a distinct sentence that simply omits the date rather than a blank or an epoch. It is defensive rather than reachable — infra's `PUT` always returns `now + cooldown` — which is fine.
- **i18n coverage is correct as-is.** Only `en.json` gains the two keys, and I checked what the other 38 locales do rather than assuming: `translateText` (`src/client/Utils.ts:454-488`) falls back to `defaultTranslations` before returning the raw key, so a French client renders the English sentence, not `account_modal.username_bare_unavailable`.
- **No injection surface from `{requested}`.** `ConfirmDialog.ts:106` renders `${this.message}` as an ordinary Lit text binding with no `unsafeHTML`, and `validateAccountUsername` has already restricted the charset before submit. Checked because it is a new user-controlled string reaching a dialog.
- **The `Api.ts:341` comment correction is accurate.** The two surviving 409s really are the public-id collision and discriminator exhaustion — that matches the endpoint as I read it reviewing infra #593, where only the claim-conflict 409 disappears.
- **The message says the right things.** It names both what the player asked for and what they got, and states that the change was spent — which is the part that actually matters, since the cooldown is the irreversible consequence.

### Suite state, including two failures I chased down

My first full client run on this branch came back **2 failed / 4173 passed (347 files)**:

- `tests/TranslationSystem.test.ts` → "en.json keys stay in sync with source usage"
- `tests/client/InventoryModal.test.ts` → "announces flag, crown and skin selections too"

The first is the one that mattered, because this PR adds two `en.json` keys and that test is precisely the key-sync guard — a real failure there would have been mine to catch. It is not one. The failure is `Error: Test timed out in 30000ms`, not an assertion, and the timeout was hiding nothing: run again with `--testTimeout=300000` it **passes in 4.52s, with the test body itself taking 2.35s**. Both tests also pass on unmodified `origin/main` in the same worktree setup.

The cause was self-inflicted — I had a full suite and an `npm ci` running concurrently when I took the first measurement. I am re-running clean and will post the number rather than leave "2 failed" standing as the last thing said about this branch.

The lesson worth recording for anyone reading CI on this repo: `TranslationSystem.test.ts` walks the source tree and sits close enough to the 30s limit under load that it fails as a timeout rather than as a key mismatch, which reads alarmingly like a real i18n regression on any PR that touches `en.json`.

---

**Verdict: correct, well-scoped, and safe to merge — the inversion you accepted was the right call and is better justified than the PR body claims.** I would take the two one-liners first: `.catch(undefined)` on the enum, and the ordering assertion. Neither blocks, but the first prevents a repeat of this exact PR's own failure mode on the next API change.

**Comment by Celant** [comment]
Full-suite follow-up, closing out the two failures I reported in my review rather than leaving "2 failed" as the last word.

Re-ran the client suite on this branch with nothing else competing: **347/347 files, 4175/4175 tests, zero failures.**

Both earlier failures were my own contention — I had a full suite and an `npm ci` running concurrently when I took the first measurement. Neither was this PR:

- `TranslationSystem.test.ts` → "en.json keys stay in sync with source usage" — the one that mattered, since this PR adds two `en.json` keys. It failed as `Error: Test timed out in 30000ms`, not an assertion. With the timeout raised it passes in **4.52s**, the test body itself taking 2.35s.
- `InventoryModal.test.ts` → "announces flag, crown and skin selections too".

Both also pass on unmodified `origin/main` in the same worktree setup.

One thing worth recording beyond this PR: `TranslationSystem.test.ts` walks the source tree and sits close enough to the 30s limit under load that it fails **as a timeout rather than as a key mismatch**. On any PR that touches `en.json` that reads exactly like a real i18n regression when it is not. Raising that file's `testTimeout` would stop it costing someone an investigation.

File: tests/client/UsernameBareClaim.test.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  updateUsername: (name: string) => updateUsername(name),
+}));
+const showInGameConfirm = vi.fn(
+  async (_message: string, _options?: unknown) => true,
+);
+const showInGameAlert = vi.fn(async (_message: string) => true);
+vi.mock("../../src/client/InGameModal", () => ({
+  showInGameConfirm: (message: string, options?: unknown) =>
+    showInGameConfirm(message, options),
+  showInGameAlert: (message: string) => showInGameAlert(message),
```
[the remark this answers is not in the document]
Comment (Celant) [reply tests/client/UsernameBareClaim.test.ts:25]: Took the first half, rejecting the second — I measured both rather than reasoning about them.

**`vi.hoisted` — done** (ceca9d7). To be precise about why, since it affects whether this was ever broken: the factories only ever *close over* the spies (`showInGameAlert: (message) => showInGameAlert(message)`), they never read them at factory time, so there was no live TDZ failure — the suite passed 10/10 before the change. But `vi.hoisted` is the idiom `tests/client/UsernameVerifiedRoute.test.ts` already uses two directories over, it costs nothing, and it removes the question permanently. Worth taking on those grounds.

**`window.location` — not changing, because the suggested alternative is the one that doesn't work here.** I probed the actual environment:

```
window.location  -> configurable: true
location.reload  -> { writable: false, enumerable: true, configurable: false }
vi.spyOn(window.location, "reload") -> threw: Cannot redefine property: reload
```

So it is the reverse of the finding: `window.location` **is** configurable in this repo's jsdom config, which is why `Object.defineProperty` works and the reload assertions genuinely fire. `reload` itself is the non-configurable one, so `vi.spyOn` — the recommended replacement — throws outright. There is no navigation boundary to use instead: the panel calls `window.location.reload()` directly and jsdom's implementation throws "not implemented", so the call has to be intercepted somewhere.

I did take the hygiene point buried in it: the descriptor is now captured up front and restored in `afterEach`, so the file no longer leaves a global replaced.

File: tests/client/UsernameBareClaim.test.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+// exist before the factories can close over them. Same shape as
+// UsernameVerifiedRoute.test.ts.
+const { updateUsername, showInGameConfirm, showInGameAlert } = vi.hoisted(
+  () => ({
+    updateUsername: vi.fn(),
+    showInGameConfirm: vi.fn(async (_message: string, _options?: unknown) =>
+      Promise.resolve(true),
+    ),
+    showInGameAlert: vi.fn(async (_message: string) => Promise.resolve(true)),
+  }),
```
[the remark this answers is not in the document]
Comment (Celant) [reply tests/client/UsernameBareClaim.test.ts:17]: Not taking this — same finding you withdrew on #5204, and it does not apply to this file.

`setup()` in `tests/util/Setup.ts` builds a game simulation from map data in `tests/testdata/maps/`. Nothing under test here touches the simulation: this is a Lit panel that PUTs to `/users/@me/username` and reads the response body. There is no game instance for `setup()` to create, and no path by which a simulation could produce a `PutUsernameResponse`.

On the mocks specifically — the three that are stubbed are the *network and the modal*, which is the point. `updateUsername` performs a `fetch`, and `showInGameAlert`/`showInGameConfirm` open a real dialog. Driving those for real would make the suite depend on a live API and a rendered modal, which is the opposite of what the CLAUDE.md line is asking for. The behaviour actually under test — the branch on `bareClaim`, and the schema parse — runs unmocked: `PutUsernameResponseSchema` is the real schema, and `UsernamePanel.handleSave` is the real method.

You reached the same conclusion on #5204 (same guidance, same kind of file) and surfaced a stored learning from PR #4670 confirming client-component tests are exempt.
